### ISSUE TITLE
Batch Transform not giving accurate results

### ISSUE THREAD

**Issue by saheelBreezo** [awslabs/amazon-sagemaker-examples#1012]
I am using this [example](https://github.com/awslabs/amazon-sagemaker-examples/blob/master/sagemaker-python-sdk/scikit_learn_inference_pipeline/Inference%20Pipeline%20with%20Scikit-learn%20and%20Linear%20Learner.ipynb) and running a custom sklearn transformer to preprocess my data however csv saved after transform is not correct . The saved csv should have more columns than the saved one. I ran this example locally and it worked fine.
Custom-transformer
```python
class CustomTransformer(BaseEstimator, TransformerMixin):
	def __init__(self,num_feature_names,cat_feature_names,redundant_features,label):
	    self.num_feature_names = num_feature_names
	    self.cat_feature_names = cat_feature_names
	    self.redundant_features = redundant_features
            self.label = label

	def fit(self,X,y=None):
	    return self

	def transform(self,X,y=None):
		X = self.filter_outlier(X)
		X = self.get_month(X)
		X = self.normalise(X)
		X = self.one_hot(X)
		X.drop(redundant_features,axis=1,inplace=True)
		return X

	def normalise(self,X):
		X_temp = X.copy()
		for feature in self.num_feature_names:
			if feature != self.label:
				max_value = X[feature].max()
				min_value = X[feature].min()
				X_temp[feature] = (X_temp[feature] - min_value) / (max_value - min_value)
		return X_temp

	def get_month(self,X):
		X.loc[:,'month'] = X['date'].dt.month
		return X

	def filter_outlier(self,X):
		X_temp = X.copy()
		for feature in self.num_feature_names:
			print(feature)
			X_temp = X_temp[((X_temp[feature] - X_temp[feature].mean()) / X_temp[feature].std()).abs() < 3]
		return X_temp
	
	def one_hot(self,X):
		X = pd.get_dummies(data=X, columns=self.cat_feature_names)
		return X
```
I am importing the above script in my preprocessing script which is exactly same as "sklearn_abalone_featurizer.py" as used in the above mentioned example and defining pipeline like this.  
`preprocessor = Pipeline(steps =[('custom_transformer',CustomTransformer(numerical_features,categorical_features,redundant_features,label))])` . Also the missing columns seems to be from one hot part.

**Comment by saheelBreezo** [comment]
Hey @mmeidl,
Thanks for the reply. The output got wrong during the **Batch transform our training data** step as per mentioned in the aws sagemaker [example](https://github.com/awslabs/amazon-sagemaker-examples/blob/master/sagemaker-python-sdk/scikit_learn_inference_pipeline/Inference%20Pipeline%20with%20Scikit-learn%20and%20Linear%20Learner.ipynb). The output of 
```python
transformer.transform(input_data,split_type='Line', content_type='text/csv')
print('Waiting for transform job: ' + transformer.latest_transform_job.job_name)
transformer.wait()
preprocessed_train = transformer.output_path
```
Now I am running the custom transformer on raw (42k,16) csv and I am getting processed (40k,33) instead of (39k,39) csv. Especially those one hot encoded columns are missing. I ran the same custom transformer locally and it worked fine.

**Comment by andremoeller** [comment]
Hey @saheelBreezo ,

You might need to modify `input_fn`, `predict_fn`, or `output_fn` to make sure columns aren't being dropped. These functions are called (in sequence) on your preprocessor model and CSV input data when you run a batch transform job. Would you be able to add logging statements to see the input data shape and return values in these functions?

**Comment by saheelBreezo** [comment]
Hey @andremoeller , thanks for reply. So this is what I found , the problem maybe more associated with  my understanding of how sklearn pipeline work. I did what you asked and kept logs. The two functions that were causing the issue were `filter_outlier` & `one_hot` . So I removed `filter_outlier` completely and replaced `one_hot` with `Pipeline(steps=[('onehot', OneHotEncoder(handle_unknown='ignore',sparse=False))])` . And it gave me the right answer.
